llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1843,12 +1844,68 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
   return false;
 }
 
+/// Memmove calls with overlapping src/dest buffers that come after a memset may
+/// be removed.
+bool MemCpyOptPass::isMemMoveMemSetDependency(MemMoveInst *M) {
+  const auto &DL = M->getDataLayout();
+  MemoryUseOrDef *MemMoveAccess = MSSA->getMemoryAccess(M);
+  if (!MemMoveAccess)
+    return false;
+
+  // The memmove is of form memmove(x, x + A, B).
+  MemoryLocation SourceLoc = MemoryLocation::getForSource(M);
+  auto *Source = dyn_cast<GetElementPtrInst>(M->getSource());
+  if (!Source)
+    return false;
+  APInt Offset(DL.getIndexTypeSizeInBits(Source->getType()), 0);
+  auto MemMoveSize = SourceLoc.Size;
+  if (!Source->accumulateConstantOffset(DL, Offset) || Offset.isNegative() ||
+      Source->getPointerOperand() != M->getDest() || !MemMoveSize.hasValue())
```

Comment:
Can move the getPointerOperand() check first, it's cheaper than accumulateConstantOffset...

Reply:
Should have done earlier.